File: Data Structures & Algorithms/max-area-of-island/submission-2.java
class Solution {
    public int maxAreaOfIsland(int[][] grid) {
        boolean[][] visited = new boolean[grid.length][grid[0].length];
        int maxArea = 0 ;
        for (int i =0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == 1 && !visited[i][j]) {
                    maxArea =  Math.max(maxArea, dfs(i, j, visited, grid));
                }
            }
        }
        return maxArea;
    }
    private int dfs(int i, int j, boolean[][] visited, int[][] grid) {
        if (i < 0 || i >= grid.length || j < 0 || j >= grid[0].length || visited[i][j] || grid[i][j] == 0 ) {
            return 0;
        }
        visited[i][j] = true;
        return 1 + dfs(i + 1, j, visited, grid) + dfs(i - 1, j, visited, grid) + dfs(i, j + 1, visited, grid) + dfs(i, j - 1, visited, grid);
    }
}

File: Data Structures & Algorithms/count-number-of-islands/submission-2.cs
public class Solution {
    public int NumIslands(char[][] grid) {
        var visited = new HashSet<(int row, int col)>();
        int numIslands = 0;
        for(int i =0; i < grid.Length; i++) {
            for(int j = 0; j < grid[0].Length; j++) {
                if (grid[i][j] == '1' && !visited.Contains((i, j))){
                    dfs(i, j, grid, visited);
                    numIslands++;
                }
            }
        }
        return numIslands;
    }
    private void dfs(int r, int c, char[][] grid, HashSet<(int row, int col)> visited) {
        if (r < 0 || r >= grid.Length || c < 0 || c >= grid[0].Length || visited.Contains((r,c)) || grid[r][c] == '0') {
            return;
        }
        visited.Add((r, c));
        dfs(r + 1, c, grid, visited);
        dfs(r - 1, c, grid, visited);
        dfs(r, c + 1, grid, visited);
        dfs(r, c - 1, grid, visited);
    }
}
